external_apps/bid/bid_action/create_bid.go

```diff
@@ -0,0 +1,273 @@
+	}
+	if activeCounterOffer != nil {
+		//7. amount needs to be less than active counter offer from owner
+		activeOfferCoin := activeCounterOffer.Amount.ToCoin(ctx.Currencies)
+		if activeOfferCoin.LessThanEqualCoin(offerCoin) {
+			return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrAmountMoreThanActiveCounterOffer, createBid.Tags(), err)
+		}
+		//8. set active counter offer to inactive
+		err = DeactivateOffer(false, bidConv.Bidder, ctx, activeCounterOffer, bidMasterStore)
+		if err != nil {
+			return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrDeactivateOffer, createBid.Tags(), err)
+		}
+	}
+	//9. lock amount
+	err = ctx.Balances.MinusFromAddress(createBid.Bidder, offerCoin)
+	if err != nil {
+		return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrLockAmount, createBid.Tags(), err)
+	}
+
+	//10. add new offer to offer store
+	createBidOffer := bid_data.NewBidOffer(
+		bidConvId,
+		bid_data.TypeBidOffer,
+		ctx.Header.Time.UTC().Unix(),
+		createBid.Amount,
+		bid_data.BidAmountLocked,
+	)
+
+	err = bidMasterStore.BidOffer.SetActiveOffer(*createBidOffer)
+	if err != nil {
+		return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrAddingOffer, createBid.Tags(), err)
+	}
+
+	return helpers.LogAndReturnTrue(ctx.Logger, createBid.Tags(), "create_bid_success")
+}
+
+func (c CreateBid) Signers() []action.Address {
+	return []action.Address{c.Bidder}
+}
+
+func (c CreateBid) Type() action.Type {
+	return BID_CREATE
+}
+
+func (c CreateBid) Tags() kv.Pairs {
+	tags := make([]kv.Pair, 0)
+
+	tag := kv.Pair{
+		Key:   []byte("tx.bidConvId"),
+		Value: []byte(c.BidConvId),
+	}
+	tag1 := kv.Pair{
+		Key:   []byte("tx.type"),
+		Value: []byte(c.Type().String()),
+	}
+	tag2 := kv.Pair{
+		Key:   []byte("tx.assetOwner"),
+		Value: c.AssetOwner.Bytes(),
+	}
+	tag3 := kv.Pair{
+		Key:   []byte("tx.asset"),
+		Value: []byte(c.AssetName),
+	}
+	tag4 := kv.Pair{
+		Key:   []byte("tx.assetType"),
+		Value: []byte(strconv.Itoa(int(c.AssetType))),
+	}
+
+	tags = append(tags, tag, tag1, tag2, tag3, tag4)
+	return tags
+}
+
+func (c CreateBid) Marshal() ([]byte, error) {
+	return json.Marshal(c)
+}
+
+func (c *CreateBid) Unmarshal(bytes []byte) error {
+	return json.Unmarshal(bytes, c)
+}
+
+func (c *CreateBid) createBidConv(ctx *action.Context) (bid_data.BidConvId, error) {
+	createBidConv := bid_data.NewBidConv(
+		c.AssetOwner,
+		c.AssetName,
+		c.AssetType,
+		c.Bidder,
+		c.Deadline,
+		ctx.Header.Height,
+	)
+	//Validate bid deadline
+	deadLine := time.Unix(createBidConv.DeadlineUTC, 0)
+
+	if deadLine.Before(ctx.Header.Time.UTC()) {
+		return "", bid_data.ErrInvalidDeadline
+	}
+
+	//Check if any bid conversation with same asset, owner, bidder already exists in active store
+	store, err := ctx.ExtStores.Get("extBidMaster")
+	if err != nil {
+		return "", bid_data.ErrGettingBidMasterStore.Wrap(err)
+	}
+	bidMasterStore := store.(*bid_data.BidMasterStore)
```

Comment:
You should handle the error ,it might lead to a panic during runtime 
bidMasterStore,ok := store.(*bid_data.BidMasterStore)

Reply:
yeah I forgot this in action pacakge.